Tutorial for nonaxisymmetric point-dipole sources in cylindrical coordinates (#2452)
NanoComp/meep: doc/docs/Python_Tutorials/Cylindrical_Coordinates.md

![](../images/zone_plate_farfield.png#center)

Nonaxisymmetric Dipole Sources
-------------------------------

In [Tutorial/Local Density of States/Extraction Efficiency of a Light-Emitting Diode (LED)](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led), the extraction efficiency of an LED was computed using an axisymmetric point-dipole source at $r = 0$. This involved a single simulation with $m = \pm 1$. Simulating a point-dipole source at $r > 0$ (as shown in the schematic below) is more challenging because it is nonaxisymmetric whereas any point source at $r > 0$ is equivalent to an axisymmetric ring source.

![](../images/cyl_nonaxisymmetric_source_layout.png#center)

> A point-dipole source at $r > 0$ can be represented as a Dirac delta function in space: $\delta(r)\delta(\phi)\delta(z) / 2\pi r$. (The $2\pi r$ factor in the denominator is necessary to ensure correct normalization.) In order to set up such a source using only axisymmetric simulations, it is necessary to expand the $\delta(\phi)$ term as a Fourier series of $\phi$: $\delta(\phi) = \frac{1}{2\pi} \sum_m e^{im\phi}$. (The Fourier transform of a Dirac delta function is a [constant](https://en.wikipedia.org/wiki/Dirac_delta_function#Fourier_transform). Each spectral component has equal weighting in its Fourier-series expansion.)

Reviewer: ```suggestion
A point-dipole source at $r_0 > 0$ can be represented as a Dirac delta function in space: $\delta(r - r_0)\delta(\phi)\delta(z) / r_0$. (The $r_0$ factor in the denominator is [necessary to ensure correct normalization](https://math.stackexchange.com/questions/398777/dirac-delta-in-polar-coordinates).) In order to set up such a source using only axisymmetric simulations, it is necessary to expand the $\delta(\phi)$ term as a Fourier series of $\phi$: $\delta(\phi) = \frac{1}{2\pi} \sum_m e^{im\phi}$. (The Fourier transform of a Dirac delta function is a [constant](https://en.wikipedia.org/wiki/Dirac_delta_function#Fourier_transform). Each spectral component has equal weighting in its Fourier-series expansion.)
```

> Simulating a point-dipole source involves two parts: (1) perform a series of simulations for $m = 0, 1, 2, ..., M$ for some cutoff $M$ of the Fourier-series expansion (the solutions for $\pm$ are simply complex conjugates, so we can omit $m <0$ and simply double the powers for $m > 0$ in the sum), and (2) because of power orthogonality, sum the results from each $m$-simulation in post processing where the $m \neq 0$ terms are multiplied by two to account for the $-m$ solutions. This procedure is described in more detail below.

Reviewer: ```suggestion
Simulating a point-dipole source involves two parts: (1) perform a series of simulations for $m = 0, 1, 2, ..., M$ for some cutoff $M$ of the Fourier-series expansion (the solutions for $\pm m$ are simply complex conjugates, so we can omit $m <0$ and simply double the powers for $m > 0$ in the sum), and (2) because of power orthogonality, sum the results from each $m$-simulation in post processing where the $m \neq 0$ terms are multiplied by two to account for the $-m$ solutions. This procedure is described in more detail below.
```

> Simulating a point-dipole source involves two parts: (1) perform a series of simulations for $m = 0, 1, 2, ..., M$ for some cutoff $M$ of the Fourier-series expansion (the solutions for $\pm m$ are simply complex conjugates, so we can omit $m <0$ and simply double the powers for $m > 0$ in the sum), and (2) because of power orthogonality, sum the results from each $m$-simulation in post processing where the $m \neq 0$ terms are multiplied by two to account for the $-m$ solutions. This procedure is described in more detail below.

Reviewer: ```suggestion
Simulating a point-dipole source involves two parts: (1) perform a series of simulations for $m = 0, 1, 2, ..., M$ for some cutoff $M$ of the Fourier-series expansion (the solutions for $\pm m$ are simply complex conjugates), and (2) because of power orthogonality, sum the results from each $m$-simulation in post processing, where the $m > 0$ terms are multiplied by two to account for the $-m$ solutions. This procedure is described in more detail below.
```

Physically, the *total* field $E(x,y,z)$ is a sum of $E_m(r,z)e^{im\phi}$ terms, one for the solution at each $m$ (similarly for $H$). Computing the total Poynting flux, however, involves integrating $\Re [E \times H^*]$ over a surface that includes an integral over $\phi$ in the range $[0,2\pi]$. The key point is that the cross terms $E_mH^*_ne^{i(m-n)\phi}$ integrate to zero due to Fourier orthogonality. **The total Poynting flux is therefore simply a sum over the Poynting fluxes calculated separately for each $m$.**

> A note regarding the source polarization at $r > 0$. The $\hat{x}$ polarization in 3d (the "in-plane" polarization) corresponds to the $\hat{r}$ polarization in cylindrical coordinates. An $\hat{r}$-polarized point-dipole source involves $\hat{r}$-polarized point sources in the $m$-simulations. Even though $\hat{r}$ is $\phi$-dependent, $\hat{r}$ is only evaluated at $\phi = 0$ because of $\delta(\phi)$. $\hat{r}$ is therefore equivalent to $\hat{x}$. This property does not hold for an $\hat{x}$-polarized point source at $r = 0$. A source at $r = 0$ does not involve $\delta(\phi)$. $\hat{x}$ is therefore defined in a more general way as $\hat{x} = \hat{r}\cos(\phi) - \hat{\phi}\sin(\phi)$. The $\sin$ and $\cos$ terms of this definition are why simulations for $m = \pm 1$ are necessary. See also [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](#scattering-cross-section-of-a-finite-dielectric-cylinder).

Reviewer: ```suggestion
A note regarding the source polarization at $r > 0$. The $\hat{x}$ polarization in 3d (the "in-plane" polarization) corresponds to the $\hat{r}$ polarization in cylindrical coordinates. An $\hat{r}$-polarized point-dipole source involves $\hat{r}$-polarized point sources in the $m$-simulations. Even though $\hat{r}$ is $\phi$-dependent, $\hat{r}$ is only evaluated at $\phi = 0$ because of $\delta(\phi)$. $\hat{r}$ is therefore equivalent to $\hat{x}$. This property does not hold for an $\hat{x}$-polarized point source at $r = 0$ (where $
\delta(\phi)$ is replaced by $1/2\pi$). A source at $r = 0$ does not involve $\delta(\phi)$. $\hat{x}$ is therefore defined in a more general way as $\hat{x} = \hat{r}\cos(\phi) - \hat{\phi}\sin(\phi)$. The $\sin$ and $\cos$ terms of this definition are why simulations for $m = \pm 1$ are necessary. See also [Tutorial/Scattering Cross Section of a Finite Dielectric Cylinder](#scattering-cross-section-of-a-finite-dielectric-cylinder).
```

Two features of this method may provide a significant speedup compared to an identical 3d simulation:

> 1. Convergence of the Fourier series may require only a small number ($M + 1$) of simulations. For a given source position $r$, $M$ can be derived analytically as $M \approx r \omega$ where $\omega$ is the angular frequency of the source within the source medium. This relationship comes from expressing the maximum phase of the fields of a *guided* mode in terms of the wavevector of a planewave in free space: $M = kr$ where $k = n\omega_0/c$ is the light line in a medium with index $n$ and $\omega_0$ is the angular frequency in vacuum ($c = 1$ in Meep units). For $m > M$, the field oscillations tend to be too rapid and the current source radiates power into non-guided (i.e., free space) modes. As an example, a point-dipole source with wavelength of $1.0 \mu m$ at a radial position of $r = 1.0 \mu m$ within a medium of $n = 2.4$ would require roughly $M = 16$ simulations. (In practice, however, we can usually truncate the Fourier-series expansion earlier without significantly degrading accuracy whenever the radiated flux at some $m$ has dropped to some small fraction of its maximum value in the summation.) The plot below shows the radiated flux vs. $m$ for three different source positions used in this tutorial example. Generally, the farther the point source is from $r = 0$, the more simulations are required for the Fourier-series summation to converge.

Reviewer: ```suggestion
1. Convergence of the Fourier series may require only a small number ($M + 1$) of simulations. For a given source position $r$, $M$ can be estimated analytically as $M \approx k r$ where $k = n\omega/c$ is the wavenumber of the source within the source medium. This comes from the fact that a source $\sim e^{im\phi}$ at $r$ oscillates in the angular direction with a spatial frequency $m/r$, but $m/r > k$ waves are evanescent, so for $m \gtrsim kr$ the radiated power tends to drop exponentially. As an example, a point-dipole source with wavelength of $1.0 \mu m$ at a radial position of $r = 1.0 \mu m$ within a medium of $n = 2.4$ would require roughly $M = 16$ simulations. (In practice, however, we can usually truncate the Fourier-series expansion earlier without significantly degrading accuracy whenever the radiated flux at some $m$ has dropped to some small fraction of its maximum value in the summation.) The plot below shows the radiated flux vs. $m$ for three different source positions used in this tutorial example. Generally, the farther the point source is from $r = 0$, the more simulations are required for the Fourier-series summation to converge.
```

            component=mp.Er,
            center=src_pt,
        ),
    ]

    geometry = [
        mp.Block(
            material=mp.Medium(index=n),
            center=mp.Vector3(0, 0, -0.5 * sz + 0.5 * dmat),
            size=mp.Vector3(mp.inf, mp.inf, dmat),
        )
    ]

    sim = mp.Simulation(
        resolution=resolution,
        cell_size=cell_size,
        dimensions=mp.CYLINDRICAL,
        m=m,
        boundary_layers=boundary_layers,
        sources=sources,
        geometry=geometry,
    )

    flux_air_mon = sim.add_flux(
        fcen,
        0,
        1,
        mp.FluxRegion(
            center=mp.Vector3(0.5 * L, 0, 0.5 * sz - dpml),
            size=mp.Vector3(L, 0, 0),
        ),
        mp.FluxRegion(
            center=mp.Vector3(L, 0, 0.5 * sz - dpml - 0.5 * dair),
            size=mp.Vector3(0, 0, dair),
        ),
    )

    sim.run(
        mp.dft_ldos(fcen, 0, 1),
        until_after_sources=mp.stop_when_fields_decayed(
            50.0,
            mp.Er,
            src_pt,
            1e-8,
        ),
    )

    flux_air = mp.get_fluxes(flux_air_mon)[0]

    if rpos == 0:
        dV = np.pi / (resolution**3)
    else:

>         dV = 2 * np.pi * rpos / (resolution**2)

Reviewer 2: Might be good to have a comment (or a note in the tutorial) that describes that this is the volume of the pixel (as briefly mentioned in the EE tutorial). However, there is an extra nuance with this approach (the pixel at r=0 is different than the rest of the grid).

Reviewer 2: (Maybe also explain where the $2\pi r$ comes from, and how to extend this to arbitrary sources (like a line).

Author: I am planning to create a dedicated page at `Features/Cylindrical Coordinates` after this PR is merged which describes this feature in more detail. This will include the Maxwell equations, Yee grid, special boundary conditions at $r = 0$, and scaling factors. This information is currently scattered in various tutorials, issues, and PRs. It would be good to collect everything into a single document.

Reviewer 2: Perfect, I'm happy to help.

Shown below is the far-field energy-density profile around the focal length for both the *r* and *z* coordinate directions for three lens designs with $N$ of 25, 50, and 100. The focus becomes sharper with increasing $N$ due to the enhanced constructive interference of the diffracted beam. As the number of zones $N$ increases, the size of the focal spot (full width at half maximum) at $z = 200$ μm decreases as $1/\sqrt{N}$ (see eq. 17 of the [reference](http://zoneplate.lbl.gov/theory)). This means that doubling the resolution (halving the spot width) requires quadrupling the number of zones.

![](../images/zone_plate_farfield.png#center)

Nonaxisymmetric Dipole Sources
-------------------------------

> In [Tutorial/Local Density of States/Extraction Efficiency of a Light-Emitting Diode (LED)](Local_Density_of_States.md#extraction-efficiency-of-a-light-emitting-diode-led), the extraction efficiency of an LED was computed using an axisymmetric point-dipole source at $r = 0$. This involved a single simulation with $m = \pm 1$. Simulating a point-dipole source at $r > 0$ (as shown in the schematic below) is more challenging because it is nonaxisymmetric: any point source at $r > 0$ is equivalent to a ring source in cylindrical coordinates. In order to simulate a point-dipole source at $r > 0$, it is necessary to represent the $\delta(r)$ (Dirac delta function) current source as a Fourier-series expansion of $\phi$ (the field angular dependence $\exp(im\phi)$). This computational approach involves two parts: (1) performing a series of simulations for $m = 0, 1, 2, ..., M$ for some cutoff $M$ of the Fourier series (described below), and (2) because of power orthogonality, summing the results from each $m$-simulation in post processing where the $m \neq 0$ terms are multiplied by two to account for the $-m$ solutions.

Reviewer: What you want is a source that is a 3d delta function:
```
δ(x)δ(y)δ(z) = δ(r)δ(φ)δ(z) / 2πr
```
and then to frame this in terms of axisymmetric simulations with exp(imφ) dependence you need to expand δ(φ) in a Fourier series
```
\delta(\phi) = \frac{1}{2\pi} \sum_m e^{im\phi}
```
up to a normalization.

Reviewer: Regarding power orthogonality:

* Physically, the total field E(x,y,z) is a sum of E_m(r,z)e^{im\phi} terms, one for the solution at each m (similarly for H).
* When you compute things like the total Poynting flux, however, you are integrating \Re [ E \times H^*] over a surface that includes an integral over φ.  In doing so, you'll see that all of the cross terms of different m's in E \times H^* integrate to zero (thanks to Fourier orthogonality), so you end up just getting a sum over the Poynting fluxes calculated separately for each m.   In short, the total power is the sum of the powers for each m.

Reviewer: Consider also the case of an x-polarized source, i.e. J proportional to `x̂`.  Then you have:
```
δ(x)δ(y)δ(z) x̂ = δ(r)δ(φ)δ(z) r̂ / 2πr
```
where the key point is that even though `r̂` is φ-dependent, because of the `δ(φ)` you are only evaluating `r̂` at φ=0 where it is the *same* as `x̂`.

This is different from point an `x̂` source at r=0, because then you write it with no `δ(φ)` as `x̂ = r̂ cos(ϕ) - ϕ̂ sin(ϕ)` and the cos and sin terms are why you do the calculation at m=1.

Author: I have expanded the discussion to address all of these points.

1. Convergence of the Fourier series may require only a small number ($M + 1$) of simulations. For a given source position $r$, $M \approx r \omega$ where $\omega$ is the angular frequency of the source within the source medium. For $m > M$, the field oscillations tend to be too rapid and the current source therefore does not radiate any power. As an example, a point dipole-source with wavelength of $1.0 \mu m$ at a radial position of $r = 1.0 \mu m$ within a medium of $n = 2.4$ would typically require 16 simulations.

2. Each $m$-simulation is independent of the others. Each $m$-simulation can therefore be executed simultaneously using an embarassingly parallel approach.

As a demonstration, we will repeat the calculation of the extraction efficiency of the LED using a point-dipole source at $r > 0$. The radiated flux of the point-dipole source (the numerator in the expression for the extraction efficiency) is computed using the Fourier--series-based method described above. A schematic of the simulation layout is shown below.

![](../images/cyl_nonaxisymmetric_source_layout.png#center)

![](../images/cyl_nonaxisymmetric_source_flux_vs_m.png#center)

> To convert the radiated flux at $r=0$ in cylindrical coordinates to 3d, the result must be multiplied by $2 / (\pi \Delta r)^2$ where $\Delta r$ is the grid size ($1/resolution$).

Reviewer: There is a factor of 1/(2πr) missing in the current source in cylindrical coordinates, because that is Jacobian factor that should go into a cylindrical delta function — Meep uses δ(r), when it should really be δ(r)/(2πr).   This is then squared in the power.   Because of the Yee grid, an Er source at r=0 is actually placed at at Yee grid point at r=Δr/2.

The factor of 2 might come from a couple of different.  To get a linearly polarized source, you need (right+left circular)/2.  So there is a factor of 2 there.      There is also probably an interpolation weight: when you ask for a source at r=0, Meep will interpolate this onto the nearest grid points, so it might put 1/2 at Δr/2 and 1/2 at –Δr/2; however, because of the mirror symmetry in r, it's possible that the latter gets combined with Δr/2 and there ends up being a weight of 1; we'd have to check.

Reviewer: Maybe add a print statement in `add_source` to see if the currents are added with weight 1 or 1/2.  (Or added twice with weight 1/2 each so that it totals to 1.)

> The relationship between the radiated flux from a point dipole at $r = 0$ ($P(0)$) and $r > 0$ ($P(r)$) is $(\pi / (\Delta r)^2) P(0) = P(r) / (\pi r)^2$.

Reviewer: There are several sources of constant factors that we would have to track down in order to get this right.  
* If there is a 1/2 factor from interpolation at Δr/2 (above) from r=0 but not at r > 0, that would give a factor of 4.
* When you sum the Fourier series m, you are including both ±m already so that changes a factor of 2 compared to above.
* In order to convert from a point source at an r=0 to a Fourier series of "circle" sources, there are some normalization factors of 1/π and 1/2π that may potentially show up.  You have to go through the algebra carefully to get these right.

However, the (1/2πr)^2 factor from the δ(r) mentioned above should be the same for both r=Δr/2 and other values of r, so this should only give you factors of (r/Δr)^2 with no pi factors.

As a demonstration, we compute the [extraction efficiency of an LED](https://meep.readthedocs.io/en/latest/Python_Tutorials/Local_Density_of_States/#extraction-efficiency-of-a-light-emitting-diode-led) from a point dipole at $r = 0$ and three different locations at $r > 0$. These results are compared to an [identical calculation in 3d](https://github.com/NanoComp/meep/blob/1fe38999997f1825054fc978e473327c77169671/python/examples/extraction_eff_ldos.py#L100-L187) for which the extraction efficiency is 0.333718. Results for the simulations in cylindrical coordinates are shown in the table below. At this resolution, the relative error is at most ~4%. The error decreases with increasing resolution.

| `rpos` | **extraction efficiency** | **relative error** |  `M` (number of `m`-simulations) |
|:------:|:-------------------------:|:------------------:|:--------------------------------:|
|    0   |          0.310405         |        0.042       |                 1                |
|   3.5  |          0.299407         |        0.041       |                56                |
|   6.7  |          0.300334         |        0.036       |                101               |

> |   9.5  |          0.302234         |        0.028       |                141               |

Reviewer 2: Oof that's a lot of `m` simulations...

Author: In this example, I intentionally used a large value (`9.5`) for the source position in $r$ (which requires a good number of individual simulations as you have noted) simply to demonstrate convincingly that the approach involving a Fourier-series expansion works even when many `m`-simulations are involved. For most practical applications, it is probably unlikely that such a large value would be used.

Also, note that it is likely that I could have probably used fewer simulations without sacrificing accuracy in the converged results. The threshold flux value used to determine when to truncate the expansion is chosen somewhat arbitrarily as `flux_tol = 1e-5` in this tutorial. We can try increasing this value.

            size=mp.Vector3(0, 0, dair),
        ),
    )

    sim.run(
        mp.dft_ldos(fcen, 0, 1),
        until_after_sources=mp.stop_when_fields_decayed(
            50.0,
            mp.Er,
            src_pt,
            1e-8,
        ),
    )

    flux_air = mp.get_fluxes(flux_air_mon)[0]

    if rpos == 0:
        dV = np.pi / (resolution**3)
    else:
        dV = 2 * np.pi * rpos / (resolution**2)

    # total flux from point source via LDOS
    flux_src = -np.real(sim.ldos_Fdata[0] * np.conj(sim.ldos_Jdata[0])) * dV

    print(f"flux-cyl:, {rpos:.2f}, {m:3d}, {flux_src:.6f}, {flux_air:.6f}")

    return flux_air, flux_src

if __name__ == "__main__":
    layer_thickness = 0.7 * wvl / n
    dipole_height = 0.5

    # r = 0 source requires a single simulation with m = ±1
    rpos = 0
    m = 1
    flux_air, flux_src = led_flux(
        layer_thickness,
        dipole_height,
        rpos,
        m,
    )
    ext_eff = flux_air / flux_src
    print(f"exteff:, {rpos}, {ext_eff:.6f}")

    # r > 0 source requires Fourier-series expansion of φ
    flux_thresh = 1e-5  # threshold value for flux for truncating expansion
    rpos = [3.5, 6.7, 9.5]
    for rp in rpos:

>         cutoff_M = int(2 * rp * 2 * np.pi * fcen * n)

Reviewer 2: We don't really mention where this comes from in the tutorial description... at the very least a comment in the code would be useful.

Author: Added some notes in the main text to describe how this expression was derived.
